### PR TITLE
tools: add coverage reporting & enforcement to router check

### LINKED ISSUE #7567: Support code coverage in router check tool
*Title*: *Enable router check tool to show code coverage for routes*

*Description*:
We are running into a situation where we want to add tests and establish a baseline coverage so that all further changes to the routes need to a test to maintain the coverage.

### PR DESCRIPTION
**Pull request by dschaller**
Description: add coverage reporting & enforcement for routes in router check tool
Risk Level: Low
Testing: Manually tested
Docs Changes: Included
Release Notes: Included
Fixes #7567
Signed-off-by: Derek Schaller <d_a_schaller@yahoo.com>

### CODE REVIEW

File: test/tools/router_check/coverage.cc [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+namespace Envoy {
+void Coverage::markCovered(const Envoy::Router::RouteEntry* route) {
+  bool seen = std::find(seen_routes_.begin(), seen_routes_.end(), route) != seen_routes_.end();
+  if (!seen) {
+    seen_routes_.push_back(route);
+  }
+}
+
+double Coverage::report() {
+  int total_route_count = 0;
```
Comment (dio) [inline test/tools/router_check/coverage.cc]: tiny nit: `size_t`?

File: docs/root/configuration/tools/router_check.rst [hunk 2]
```diff
@@ -3,9 +3,11 @@
 Route table check tool
 ======================
 
-**NOTE: The following configuration is for the route table check tool only and is not part of the Envoy binary.
-The route table check tool is a standalone binary that can be used to verify Envoy's routing for a given configuration
-file.**
+.. note::
```
Comment (dio) [inline docs/root/configuration/tools/router_check.rst]: Thank you for this.

**Comment by dschaller** [comment]
@dio thanks for the review! Updated based on your feedback

File: test/tools/router_check/router.cc [hunk 3]
```diff
@@ -281,7 +282,11 @@ bool RouterCheckTool::compareRewritePath(ToolConfig& tool_config, const std::str
 
     actual = tool_config.headers_->get_(Http::Headers::get().Path);
   }
-  return compareResults(actual, expected, "path_rewrite");
+  bool matches = compareResults(actual, expected, "path_rewrite");
```
Comment (rgs1) [inline test/tools/router_check/router.cc]: nit: `const bool ...`

File: test/tools/router_check/router_check.cc [hunk 4]
```diff
@@ -23,6 +24,15 @@ int main(int argc, char* argv[]) {
     if (!is_equal) {
       return EXIT_FAILURE;
     }
+
+    double current_coverage = checktool.coverage();
```
Comment (rgs1) [inline test/tools/router_check/router_check.cc]: nit: `const double`

File: test/tools/router_check/coverage.cc [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#include "test/tools/router_check/coverage.h"
+
+#include <algorithm>
+
+#include "envoy/api/v2/core/base.pb.h"
+
+namespace Envoy {
+void Coverage::markCovered(const Envoy::Router::RouteEntry* route) {
+  bool seen = std::find(seen_routes_.begin(), seen_routes_.end(), route) != seen_routes_.end();
```
Comment (rgs1) [inline test/tools/router_check/coverage.cc]: nit: `const bool`

**Comment by dschaller** [comment]
@mattklein123 thanks! Definitely agree about the follow up. I wanted to get basic coverage reporting and enforcement but I’d like to expand this to produce reports and include more comprehensive coverage of the fields for each route entry

**Comment by mattklein123** [comment]
> @mattklein123 thanks! Definitely agree about the follow up. I wanted to get basic coverage reporting and enforcement but I’d like to expand this to produce reports and include more comprehensive coverage of the fields for each route entry

Awesome!

**Comment by dschaller** [comment]
@dio @rgs1 @mattklein123 updated this based on feedback. Can I get another look over this? :)

File: test/tools/router_check/coverage.cc [hunk 6]
```diff
@@ -0,0 +1,23 @@
+#include "test/tools/router_check/coverage.h"
+
+#include <algorithm>
+
+#include "envoy/api/v2/core/base.pb.h"
+
+namespace Envoy {
+void Coverage::markCovered(const Envoy::Router::RouteEntry* route) {
+  const bool seen =
+      std::find(seen_routes_.begin(), seen_routes_.end(), route) != seen_routes_.end();
```
Comment (mattklein123) [inline test/tools/router_check/coverage.cc]: nit: just fold this into the if statement below

File: test/tools/router_check/coverage.cc [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  if (!seen) {
+    seen_routes_.push_back(route);
+  }
+}
+
+double Coverage::report() {
+  int size_t = 0;
```
Comment (mattklein123) [inline test/tools/router_check/coverage.cc]: nit: `uint64_t num_routes`

File: test/tools/router_check/coverage.h [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+namespace Envoy {
+class Coverage : Logger::Loggable<Logger::Id::testing> {
+public:
+  Coverage(envoy::api::v2::RouteConfiguration config) : route_config_(config){};
+  void markCovered(const Envoy::Router::RouteEntry* route);
+  double report();
+
+private:
+  std::vector<const Envoy::Router::RouteEntry*> seen_routes_;
+  envoy::api::v2::RouteConfiguration route_config_;
```
Comment (mattklein123) [inline test/tools/router_check/coverage.h]: const

File: test/tools/router_check/coverage.h [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+#include "common/stats/fake_symbol_table_impl.h"
+
+#include "test/mocks/server/mocks.h"
+#include "test/test_common/utility.h"
+
+namespace Envoy {
+class Coverage : Logger::Loggable<Logger::Id::testing> {
+public:
+  Coverage(envoy::api::v2::RouteConfiguration config) : route_config_(config){};
+  void markCovered(const Envoy::Router::RouteEntry* route);
```
Comment (mattklein123) [inline test/tools/router_check/coverage.h]: Pass a const reference as this can't be null

File: test/tools/router_check/router.cc [hunk 10]
```diff
@@ -73,16 +73,17 @@ RouterCheckTool RouterCheckTool::create(const std::string& router_config_file) {
   auto factory_context = std::make_unique<NiceMock<Server::Configuration::MockFactoryContext>>();
   auto config = std::make_unique<Router::ConfigImpl>(route_config, *factory_context, false);
 
+  Coverage coverage_report = Coverage(route_config);
```
Comment (mattklein123) [inline test/tools/router_check/router.cc]: nit: remove local variable, pass inline below

**Comment by jyotimahapatra** [comment]
#7567

**Comment by jyotimahapatra** [comment]
We possibly can add a small check in route_tests.sh

File: test/tools/router_check/coverage.h (hunk 9 above)
Comment (dschaller) [reply test/tools/router_check/coverage.h]: I might be missing something but if I use a reference here won't I have comparison issues? I've attempt to implement this and haven't been successful but it's very likely that I'm overlooking something simple.

Comment (mattklein123) [reply test/tools/router_check/coverage.h]: A reference is just a pointer that isn't null. If you want to compare a reference to a pointer all you need to do is take the address of it (`ptr == &ref`).

Comment (dschaller) [reply test/tools/router_check/coverage.h]: Ah ok. I think I actually found my issue. Simple oversight on my part :)

File: test/tools/router_check/coverage.cc [hunk 11]
```diff
@@ -0,0 +1,21 @@
+#include "test/tools/router_check/coverage.h"
+
+#include <algorithm>
+
+#include "envoy/api/v2/core/base.pb.h"
+
+namespace Envoy {
+void Coverage::markCovered(const Envoy::Router::RouteEntry& route) {
+  if (std::find(seen_routes_.begin(), seen_routes_.end(), &route) == seen_routes_.end()) {
```
Comment (mattklein123) [inline test/tools/router_check/coverage.cc]: I think this is the inverse of what you want? If so I think your tests are not working. :)

Comment (jyotimahapatra) [reply test/tools/router_check/coverage.cc]: using std::unordered_set can possibly simplify it?
like this? https://github.com/jyotima/envoy/blob/7690b8ddb0927715840ffe4b0919bcfdf6174762/test/tools/router_check/router.cc#L173

Comment (dschaller) [reply test/tools/router_check/coverage.cc]: I think this is right. If find reaches the end of the already seen routes then we know we haven’t encountered this route before. If it doesn’t then we skip adding it as a duplicate.

I can also look at using an unordered set if that’s preferred.

Comment (mattklein123) [reply test/tools/router_check/coverage.cc]: Oops sorry, my bad. Can you add a small comment?

File: test/tools/router_check/router_check.cc [hunk 12]
```diff
@@ -7,6 +7,7 @@
 int main(int argc, char* argv[]) {
   Envoy::Options options(argc, argv);
 
+  const bool enforceCoverage = options.failUnder() != 0.0;
```
Comment (mattklein123) [inline test/tools/router_check/router_check.cc]: nit: `enforce_coverage`

**Comment by mattklein123** [comment]
@dio any additional comments?

**Comment by jyotimahapatra** [comment]
#730

**Pull request merged**